File: programs/dippies_index_protocol/src/instructions/replace_node.rs
use anchor_lang::prelude::*;

use crate::constants::{FOREST_SEED, MAX_CHILD_PER_NODE, NODE_SEED, TREE_SEED};
use crate::errors::DipErrors;
use crate::state::{Forest, Node, Tree};

pub fn replace_node(ctx: Context<ReplaceNode>) -> Result<()> {
    msg!("Replacing a node");

    let parent_node = &mut ctx.accounts.parent_node;
    let position = parent_node
        .children
        .iter()
        .position(|&n| n == ctx.accounts.weaker_node.key())
        .unwrap();
    parent_node.children.remove(position);
    parent_node.children.push(ctx.accounts.node.key());

    Ok(())
}

#[derive(Accounts)]
pub struct ReplaceNode<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    /// The forest
    #[account(
        seeds = [
            FOREST_SEED.as_bytes(),
            &forest.id.to_bytes(),
        ],
        bump,
    )]
    pub forest: Account<'info, Forest>,

    /// The tree
    #[account(
        seeds = [
            TREE_SEED.as_bytes(),
            &forest.key().to_bytes(),
            &tree.title.as_ref(),
        ],
        bump,
    )]
    pub tree: Account<'info, Tree>,

    /// The parent node to attach to
    #[account(
        mut,
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &parent_node.parent.key().to_bytes(),
            &parent_node.tags.last().unwrap().as_ref(),
        ],
        bump,
        constraint = parent_node.children.len() == MAX_CHILD_PER_NODE @ DipErrors::NodeNotFull,
        constraint = parent_node.children.iter().find(|&n| n == &node.key()).is_none() @ DipErrors::AlreadyAChild,
        constraint = parent_node.children.iter().find(|&n| n == &weaker_node.key()).is_some() @ DipErrors::NotAChild,
    )]
    pub parent_node: Account<'info, Node>,

    /// The attached node
    #[account(
        mut,
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &parent_node.key().to_bytes(),
            &node.tags.last().unwrap().as_ref(),
        ],
        bump,
        constraint = node.stake > weaker_node.stake @ DipErrors::NotEnoughStake,
    )]
    pub node: Account<'info, Node>,

    /// The weaker node being replaced
    #[account(
        mut,
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &parent_node.key().to_bytes(),
            &weaker_node.tags.last().unwrap().as_ref(),
        ],
        bump,
    )]
    pub weaker_node: Account<'info, Node>,

    /// Common Solana programs
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}

File: ts/instructions/replaceNode.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface ReplaceNodeAccounts {
  signer: PublicKey
  /** The forest */
  forest: PublicKey
  /** The tree */
  tree: PublicKey
  /** The parent node to attach to */
  parentNode: PublicKey
  /** The attached node */
  node: PublicKey
  /** The weaker node being replaced */
  weakerNode: PublicKey
  /** Common Solana programs */
  systemProgram: PublicKey
  rent: PublicKey
}

export function replaceNode(accounts: ReplaceNodeAccounts) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.forest, isSigner: false, isWritable: false },
    { pubkey: accounts.tree, isSigner: false, isWritable: false },
    { pubkey: accounts.parentNode, isSigner: false, isWritable: true },
    { pubkey: accounts.node, isSigner: false, isWritable: true },
    { pubkey: accounts.weakerNode, isSigner: false, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.rent, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([0, 166, 208, 185, 177, 131, 244, 19])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}
